File: src/main/java/org/mechaflow/standard/ports/SignalPort.java
package org.mechaflow.standard.ports;

import org.flowutils.time.Time;
import org.mechaflow.PortBase;
import org.mechaflow.PortDirection;

/**
 * Port that propagates signals from input to output ports.
 */
public final class SignalPort extends PortBase {

    public static final float HIGH_LEVEL = 0.5f;

    private float value;

    public SignalPort(String name, PortDirection direction) {
        this(name, direction, null);
    }

    public SignalPort(String name, PortDirection direction, String description) {
        this(name, direction, description, DEFAULT_SMALL_SIZE_GAUGE);
    }

    public SignalPort(String name, PortDirection direction, float defaultValue, String description) {
        this(name, direction, description, DEFAULT_SMALL_SIZE_GAUGE);
        set(defaultValue);
    }

    public SignalPort(String name, PortDirection direction, String description, int sizeGauge) {
        this(name, direction, description, sizeGauge, sizeGauge);
    }

    public SignalPort(String name, PortDirection direction, String description, int widthGauge, int heightGauge) {
        super(name, direction, description, widthGauge, heightGauge);
        if (direction == PortDirection.INOUT) throw new IllegalArgumentException("A signal port must be either an input or output port, INOUT signal ports are not supported");
    }

    @Override public void propagate(Time time) {
        // Propagate our signal to the connected port
        if (getDirection() == PortDirection.OUT && isConnected()) {
            ((SignalPort) getConnectedPort()).set(value);
        }
    }

    public float get() {
        return value;
    }

    public void set(float value) {
        this.value = value;
    }

    public final boolean isHigh() {
        return value >= HIGH_LEVEL;
    }

    public final boolean isLow() {
        return value < HIGH_LEVEL;
    }

}

File: src/main/kotlin/org/mechaflow2/standard/ports/SignalPort.kt
package org.mechaflow2.standard.ports

import org.flowutils.time.Time
import org.mechaflow2.PortBase
import org.mechaflow2.PortDirection
import org.mechaflow2.PortSize


/**
 * Port that propagates signals from input to output ports.
 */
class SignalPort(name: String,
                 direction: PortDirection = PortDirection.IN,
                 val defaultValue: Double = 0.0,
                 val range: SignalRange = SignalRange.UNLIMITED,
                 description: String? = null,
                 size: PortSize = PortSize.MEDIUM) :
        PortBase(name, direction, description, size) {

    var value: Double = defaultValue
        set(newValue) {
            value = range.clamp(newValue)
        }

    init {
        if (direction == PortDirection.INOUT) {
            throw IllegalArgumentException(
                    "A signal port must be either an input or output port, INOUT signal ports are not supported")
        }
    }

    override fun propagate(time: Time) {
        // Propagate our signal to the connected port
        if (direction == PortDirection.OUT && isConnected()) {
            (connectedPort as SignalPort).value = this.value
        }
    }

    val isHigh: Boolean
        get() = value >= HIGH_LEVEL

    val isLow: Boolean
        get() = value < HIGH_LEVEL

    companion object {
        const val HIGH_LEVEL = 0.5
    }

}
